pallets/pips/src/lib.rs

```diff
@@ -1600,6 +1484,79 @@ impl<T: Trait> Module<T> {
         Ok(())
     }
 
+    /// Construct a `SnapshottedPip` from a `PipId`.
+    /// `true` denotes a positive sign.
+    fn aggregate_result(id: PipId) -> SnapshottedPip<BalanceOf<T>> {
+        let VotingResult {
+            ayes_stake,
+            nays_stake,
+            ..
+        } = <ProposalResult<T>>::get(id);
+        let weight = if ayes_stake >= nays_stake {
+            (true, ayes_stake - nays_stake)
+        } else {
+            (false, nays_stake - ayes_stake)
+        };
+        SnapshottedPip { id, weight }
+    }
+
+    /// Adjust the live queue under the assumption that `id` should be moved up or down the queue.
+    fn adjust_live_queue(id: PipId, old: SnapshottedPip<BalanceOf<T>>) {
+        let new = Self::aggregate_result(id);
+        <LiveQueue<T>>::mutate(|queue| {
+            let old_pos = queue.binary_search_by(|res| compare_spip(res, &old));
+            let new_pos = queue.binary_search_by(|res| compare_spip(res, &new));
+            match (old_pos, new_pos) {
+                // First time adding to queue, so insert.
+                (Err(_), Ok(pos) | Err(pos)) => queue.insert(pos, new),
+                // Already in queue. Swap positions.
+                (Ok(old_pos), Err(new_pos)) => {
+                    // Cannot underflow by definition of binary search finding an element.
+                    let new_pos = new_pos.min(queue.len() - 1);
+                    move_element(queue, old_pos, new_pos);
+                    queue[new_pos] = new;
+                }
+                // We have `old_res == new_res`. Queue state is already good.
+                (Ok(_), Ok(_)) => {}
+            }
+        });
+    }
+
+    /// Insert a new PIP into the live queue.
+    ///
+    /// The `id` should not exist in the queue previously.
+    /// Panics if it did.
+    fn insert_live_queue(id: PipId) {
+        let new = Self::aggregate_result(id);
+        <LiveQueue<T>>::mutate(|queue| {
+            // Inserting a new PIP entails that `id` is nowhere to be found.
+            // It follows that binary search will return `Err(_)`.
+            let pos = queue
+                .binary_search_by(|res| compare_spip(res, &new))
+                .unwrap_err();
```

Comment:
It would be great to add a unit test that ensures to some degree that `Ok` is not returned here. It could be tested as part of `propose` but an explicit test would be more ascertaining, I think.

Reply:
Per private slack convo a positive test provoking a panic (by manual queue manipulation) will be added instead.